build-conventions/src/main/java/org/elasticsearch/gradle/internal/conventions/precommit/LicenseHeadersTask.java

```diff
@@ -0,0 +1,271 @@
+            unapprovedFiles(getReportFile()).stream().forEachOrdered(unapprovedFile -> getLogger().error(unapprovedFile));
+            throw new GradleException("Check failed. License header problems were found. Full details: " + reportFile.getAbsolutePath());
+        }
+    }
+
+    private IHeaderMatcher subStringMatcher(String licenseFamilyCategory, String licenseFamilyName, String substringPattern) {
+        SubstringLicenseMatcher substringLicenseMatcher = new SubstringLicenseMatcher();
+        substringLicenseMatcher.setLicenseFamilyCategory(licenseFamilyCategory);
+        substringLicenseMatcher.setLicenseFamilyName(licenseFamilyName);
+        SubstringLicenseMatcher.Pattern pattern = new SubstringLicenseMatcher.Pattern();
+        pattern.setSubstring(substringPattern);
+        substringLicenseMatcher.addConfiguredPattern(pattern);
+        return substringLicenseMatcher;
+    }
+
+    private ClaimStatistic generateReport(ReportConfiguration config, File xmlReportFile) {
+        try {
+            Files.deleteIfExists(reportFile.toPath());
+            BufferedWriter bufferedWriter = new BufferedWriter(new FileWriter(xmlReportFile));
+            return toXmlReportFile(config, bufferedWriter);
+        } catch (IOException | RatException exception) {
+            throw new GradleException("Cannot generate license header report for " + getPath(), exception);
+        }
+    }
+
+    private ClaimStatistic toXmlReportFile(ReportConfiguration config, Writer writer) throws RatException, IOException {
+        ClaimStatistic stats = new ClaimStatistic();
+        RatReport standardReport = XmlReportFactory.createStandardReport(new XmlWriter(writer), stats, config);
+
+        standardReport.startReport();
+        for (FileCollection dirSet : getSourceFolders().get()) {
+            for (File f : dirSet.getAsFileTree().matching(patternFilterable -> patternFilterable.exclude(getExcludes()))) {
+                standardReport.report(new FileDocument(f));
+            }
+        }
+        standardReport.endReport();
+        writer.flush();
+        writer.close();
+        return stats;
+    }
+
+    private static List<String> unapprovedFiles(File xmlReportFile) {
+        try {
+            NodeList resourcesNodes = DocumentBuilderFactory.newInstance()
+                .newDocumentBuilder()
+                .parse(xmlReportFile)
+                .getElementsByTagName("resource");
+            return elementList(resourcesNodes).stream()
+                .filter(
+                    resource -> elementList(resource.getChildNodes()).stream()
+                        .anyMatch(n -> n.getTagName().equals("license-approval") && n.getAttribute("name").equals("false"))
+                )
+                .map(e -> e.getAttribute("name"))
+                .sorted()
+                .collect(Collectors.toList());
+        } catch (SAXException | IOException | ParserConfigurationException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)